_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,352 @@
+$ kafka-cp my-group
+12
+
+# transaction states are in partition 30 of __transaction_state
+$ kafka-cp kafka-txn-0
+30
+```
+
+Knowing the `output-topic` and coordinating partitions, we can take a look at their content using the embedded dump tool.
+Note how we pass the decoder parameter when dumping from internal topics, whose content is encoded for performance reasons.
+
+```sh
+ 1 # dump output-topic-0
+ 2 $ $KAFKA_HOME/bin/kafka-dump-log.sh --deep-iteration --print-data-log --files /tmp/kafka-logs/output-topic-0/00000000000000000000.log
+ 3 Dumping /tmp/kafka-logs/output-topic-0/00000000000000000000.log
+ 4 Log starting offset: 0
+ 5 baseOffset: 0 lastOffset: 0 count: 1 baseSequence: 0 lastSequence: 0 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: false deleteHorizonMs: OptionalLong.empty position: 0 CreateTime: 1680383687941 size: 82 magic: 2 compresscodec: none crc: 2785707995 isvalid: true
+ 6 | offset: 0 CreateTime: 1680383687941 keySize: -1 valueSize: 14 sequence: 0 headerKeys: [] payload: tset
+ 7 baseOffset: 1 lastOffset: 1 count: 1 baseSequence: -1 lastSequence: -1 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: true deleteHorizonMs: OptionalLong.empty position: 82 CreateTime: 1680383688163 size: 78 magic: 2 compresscodec: none crc: 3360473936 isvalid: true
+ 8 | offset: 1 CreateTime: 1680383688163 keySize: 4 valueSize: 6 sequence: -1 headerKeys: [] endTxnMarker: COMMIT coordinatorEpoch: 0
+ 9 
+10 # dump __consumer_offsets-12 using the offsets decoder
+11 $ $KAFKA_HOME/bin/kafka-dump-log.sh --deep-iteration --print-data-log --offsets-decoder --files /tmp/kafka-logs/__consumer_offsets-12/00000000000000000000.log
+12 Dumping /tmp/kafka-logs/__consumer_offsets-12/00000000000000000000.log
+13 Starting offset: 0
+14 ...
+15 baseOffset: 1 lastOffset: 1 count: 1 baseSequence: 0 lastSequence: 0 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: false deleteHorizonMs: OptionalLong.empty position: 339 CreateTime: 1665506597950 size: 118 magic: 2 compresscodec: none crc: 4199759988 isvalid: true
+16 | offset: 1 CreateTime: 1680383688085 keySize: 26 valueSize: 24 sequence: 0 headerKeys: [] key: offset_commit::group=my-group,partition=input-topic-0 payload: offset=1
+17 baseOffset: 2 lastOffset: 2 count: 1 baseSequence: -1 lastSequence: -1 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: true deleteHorizonMs: OptionalLong.empty position: 457 CreateTime: 1665506597998 size: 78 magic: 2 compresscodec: none crc: 3355926470 isvalid: true
+18 | offset: 2 CreateTime: 1665506597998 keySize: 4 valueSize: 6 sequence: -1 headerKeys: [] endTxnMarker: COMMIT coordinatorEpoch: 0
+19 ...
+```
+
+The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).
+In `wc-output-0` dump, we see that the data batch is transactional and contains the producer's state (line 5).
```

Comment:
```suggestion
In `output-topic-0` dump, we see that the data batch is transactional and contains the producer's state (line 5).
```

Reply:
also where is the producer's state?

Reply:
I meant producer session, as defined above. Fixed.